backend/src/main/kotlin/cz/loono/backend/api/service/HealthcareProvidersService.kt

```diff
@@ -1,40 +1,143 @@
 import cz.loono.backend.data.constants.Constants.OPEN_DATA_URL
 import cz.loono.backend.db.model.HealthcareCategory
+import cz.loono.backend.db.model.HealthcareProvider
 import cz.loono.backend.db.repository.HealthcareCategoryRepository
 import cz.loono.backend.db.repository.HealthcareProviderRepository
+import io.github.reactivecircus.cache4k.Cache
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.http.HttpStatus
+import org.springframework.scheduling.annotation.Scheduled
 import org.springframework.stereotype.Service
 import org.springframework.transaction.annotation.Transactional
+import java.io.ByteArrayOutputStream
+import java.io.IOException
 import java.net.URL
+import java.util.zip.ZipEntry
+import java.util.zip.ZipOutputStream
 
 @Service
 class HealthcareProvidersService @Autowired constructor(
     private val healthcareProviderRepository: HealthcareProviderRepository,
     private val healthcareCategoryRepository: HealthcareCategoryRepository
 ) {
 
+    private val providersCache = Cache.Builder().build<HealthcareProviderIdDto, HealthcareProvider>()
+    private val fileCache = Cache.Builder().build<String, ByteArray>()
+
+    @Scheduled(cron = "0 0 2 1 * ?") // each the 1st day of month at 2AM
+    @Synchronized
     @Transactional(rollbackFor = [Exception::class])
-    fun updateData(): UpdateStatusMessage {
+    fun updateData(): UpdateStatusMessageDto {
         val input = URL(OPEN_DATA_URL).openStream()
         val providers = HealthcareCSVParser().parse(input)
         if (providers.isNotEmpty()) {
             val categoryValues = CategoryValues.values().map { HealthcareCategory(value = it.value) }
             healthcareCategoryRepository.saveAll(categoryValues)
             healthcareProviderRepository.saveAll(providers)
+            updateCache()
         } else {
             throw LoonoBackendException(
                 HttpStatus.UNPROCESSABLE_ENTITY,
                 errorCode = HttpStatus.UNPROCESSABLE_ENTITY.value().toString(),
                 errorMessage = "Data update failed."
             )
         }
-        return UpdateStatusMessage("Data successfully updated.")
+        return UpdateStatusMessageDto("Data successfully updated.")
+    }
+
+    private fun updateCache() {
+        providersCache.invalidateAll()
+        fileCache.invalidate("providers")
+        val providers = healthcareProviderRepository.findAll()
+        providers.forEach { provider ->
+            val id = HealthcareProviderIdDto(locationId = provider.locationId, institutionId = provider.institutionId)
+            providersCache.put(id, provider)
+        }
+        fileCache.put("providers", zipProviders())
+    }
+
+    private fun zipProviders(): ByteArray {
+        val simplifyProviders = providersCache.asMap().values.map { it.simplify() }
+        val list = HealthcareProviderListDto(
+            healthcareProviders = simplifyProviders
+        )
+        val `in` = Gson().toJson(list)
```

Comment:
Musí sa tá premenná volať tak divne, aby tam boli ``? Príde mi to ťažko čiteľné. ako by si tam robil niečo špeciálne.